### DISCUSSION TITLE
Running concurrently asyncio.start_server() function inside cocotb testbench

### DISCUSSION THREAD

**Discussion by RapidFo** [cocotb/cocotb#3037] [Q&A]
Is there a way to run concurrently asyncio.start_server() function inside cocotb testbench? Code similar to [TCP echo server using streams](https://docs.python.org/3/library/asyncio-stream.html#tcp-echo-server-using-streams) example.

Inside handle_echo I want to read / write four jtag signals using OpenOCD's bitbang, so I need TCP server inside cocotb testbench. In same time I want to read / write other signals. 

When I run TCP server with asyncio.run(main()) inside my test, then only TCP server works. I tried cocotb.external but I got milion errors...

**Comment by marlonjames** [comment]
You could probably run it in a separate thread, but the main issue is realizing that the cocotb test thread is at the mercy of the simulator. Unless a Trigger fires, no cocotb code will run, and there is no standardized way to interrupt a running simulation.
